### PR TITLE
no-curly-component-invocation: Fix false positives

### PR DESCRIPTION
**Pull request by Turbo87**
This PR resolves https://github.com/ember-template-lint/ember-template-lint/issues/1026 by basically rewriting the rule 🙈 

I've documented the assumptions that the rule now makes in the rule documentation.

I would personally consider this a bugfix since it clarifies some edge cases and should result in less false positive warnings.

### CODE REVIEW

File: test/unit/rules/no-curly-component-invocation-test.js [hunk 1]
```diff
@@ -6,71 +6,177 @@ function generateError(name) {
   return `You are using the component {{${name}}} with curly component syntax. You should use <${angleBracketName}> instead. If it is actually a helper you must manually add it to the 'no-curly-component-invocation' rule configuration, e.g. \`'no-curly-component-invocation': { allow: ['${name}'] }\`.`;
 }
 
+const SHARED_GOOD = [
+  '{{foo}}',
```
Comment (rwjblue) [inline test/unit/rules/no-curly-component-invocation-test.js]: This should not be a "good" example. The octane configuration _already_ requires the template to avoid ambiguous paths making `{{foo}}` reserved for an invocation (either a helper or a component). When `foo` is not a local, it should print an error message that is a little different than the more obvious cases (e.g. it should mention that the path is ambiguous and they should either refactor to `{{this.foo}}` or `<Foo />`).

File: lib/rules/no-curly-component-invocation.js [hunk 2]
```diff
@@ -78,89 +46,132 @@ module.exports = class NoCurlyComponentInvocation extends Rule {
...
+
+        let hasNamedArguments = node.hash.pairs.length !== 0;
+        if (!hasNamedArguments) {
+          if (path.parts.length > 1) {
+            // {{foo.bar}}
+            return;
+          }
+
+          let name = path.original;
+          if (this.config.allow && this.config.allow.includes(name)) {
```
[the remark this answers is not in the document]
Comment (Turbo87) [reply lib/rules/no-curly-component-invocation.js]: yeah, totally agree. I just used what was there without touching the `parseConfig()` method for now, but I can adjust that too

[the remark this answers is not in the document]
Comment (rwjblue) [reply lib/rules/no-curly-component-invocation.js]: Hehe, you rewrote everything else so....

File: test/unit/rules/no-curly-component-invocation-test.js (hunk 1 above)
Comment (Turbo87) [reply test/unit/rules/no-curly-component-invocation-test.js]: This means we assume `{{foo}}` is a component invocation, not a helper invocation?

IMHO that should be configurable then. I've seen enough cases where apps would have been able to use the `no-curly-component-invocation` rule, but haven't migrated to `no-implicit-this` yet 🤔

Comment (Turbo87) [reply test/unit/rules/no-curly-component-invocation-test.js]: also, see https://github.com/ember-template-lint/ember-template-lint/issues/1026 😉

Comment (rwjblue) [reply test/unit/rules/no-curly-component-invocation-test.js]: Yes, I have seen #1026 and replied in detail there.

Comment (rwjblue) [reply test/unit/rules/no-curly-component-invocation-test.js]: > IMHO that should be configurable then. 

Yep, sounds totally reasonable.

Comment (rwjblue) [reply test/unit/rules/no-curly-component-invocation-test.js]: We can then make it default to making no assumptions about `{{foo}}` _if_ `no-implicit-this` is disabled.

File: lib/rules/no-curly-component-invocation.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (Turbo87) [reply lib/rules/no-curly-component-invocation.js]: yeeeeaaaahh..... 🙈 

😆

File: test/unit/rules/no-curly-component-invocation-test.js (hunk 1 above)
Comment (Turbo87) [reply test/unit/rules/no-curly-component-invocation-test.js]: done!

**Comment by rwjblue** [comment]
Thank you @Turbo87!

File: docs/rule/no-curly-component-invocation.md [hunk 3]
```diff
@@ -1,115 +1,92 @@
...
+Edition.
+
+This rule checks all the curly braces in your app and warns about those that
+look like they could be component invocations.
+
+### Examples
+
+- `{{foo}}` ✅
+  (simple mustache without `-` in the path is likely to be a property; unless
+  `noImplicitThis` is set or in `disallow` list and not a scoped variable)
```
Comment (chancancode) [inline docs/rule/no-curly-component-invocation.md]: Do we check for local variables when `noImplicitThis` is set? (`{{#let ... as |foo|}}{{foo}}{{/let}}` should not error.)

File: docs/rule/no-curly-component-invocation.md [hunk 4]
```diff
@@ -1,115 +1,92 @@
...
+
+- `{{foo.bar}}` ✅
+  (simple mustache with nested path is likely to be a nested property; unless
+  `noImplicitThis` is set)
```
Comment (chancancode) [inline docs/rule/no-curly-component-invocation.md]: same question re:above

File: docs/rule/no-curly-component-invocation.md [hunk 5]
```diff
@@ -1,115 +1,92 @@
...
+
+- `{{foo-bar}}` ❌ (angle brackets: `<FooBar />`)
+  (simple mustache with `-` in the path is more likely a component than a
+  property or helper; unless in `allow` list)
+
+- `{{nested/component}}` ❌ (angle brackets: `<Nested::Component />`)
+  (simple mustache with `/` in the path is more likely a component than a
+  property or helper; unless in `allow` list)
+
+- `{{42}}` ✅
```
Comment (chancancode) [inline docs/rule/no-curly-component-invocation.md]: I assume this covers `{{true}}`, `{{false}}`, `{{null}}` and `{{undefined}}` right?

**Comment by chancancode** [comment]
Regression: this PR is currently rejecting `{{outlet}}` due to removal of the built-in list, @Turbo87 is looking into it! See also https://discordapp.com/channels/480462759797063690/514504648590098457/658796846881177601 in discord.

File: docs/rule/no-curly-component-invocation.md (hunk 3 above)
Comment (Turbo87) [reply docs/rule/no-curly-component-invocation.md]: yes, we do

File: docs/rule/no-curly-component-invocation.md (hunk 5 above)
Comment (Turbo87) [reply docs/rule/no-curly-component-invocation.md]: yes

File: docs/rule/no-curly-component-invocation.md (hunk 4 above)
Comment (Turbo87) [reply docs/rule/no-curly-component-invocation.md]: apparently not yet, but I'll fix that too
